Introduce pubtools-ami-delete command [RHELDST-7515]

rbikar opened the pull request:
This commit introduces pubtools-ami-delete command that
enables removal of AMI images and snapshot from AWS + it
does related update in RHSM metadata services (makes image
invisible). Deletion in AWS is done via cloudimg library
that communicates directly with AWS APIs.

There was rather large refactoring done, most notably:
1) Created `AmiBase` class and moved there common logic
   and args for both `push` and `delete` commands.
2) Moved exception classes to their own file exceptions.py.
3) Added saving/updating push items via Collector
4) Added usage of `step` decorator.

rbikar commented:
Notes:

1) Little deviation from specification in JIRA in cmd args (cc @MichalHaluza ) :

-- using singular `--keep-snapshot` insted `--keep-snapshots`
-- not introduced `--concurrent`, this can be set via env var: AMI_DELETE_REQUEST_THREADS
-- implemented `--source` as positional arg, as it is in pubtools-ami-push

2) context manager for executors seems to be broken whhen using composing syntax

This works:
```
with Executors.thread_pool() as ex:
   f = ex.submit()
   f.result()
   
```

this does seem to hang indefinitely:
```
with Executors.thread_pool().with_retry() as ex:
   f = ex.submit()
   f.result()
```

rohanpm commented on tests/push/test_ami_delete.py:
```diff
@@ -0,0 +1,324 @@
+    )
+    source = "pub:https://pub.example.com?task_id=1234"
+
+    requests_mocker.register_uri(
```
Personally I don't think this is the right way to set up the test data here.  The test has been tied to a specific implementation detail of how pushsource library obtains data for 'pub' backend when it was unnecessary to do so.  If that implementation detail changes then the test will start to fail.  It's not the job of pubtools-ami's test suite to test the pushsource pub backend.

If the pubtools-ami code can work with any pushsource backend, then I think the better approaches here are:

- use 'staged' backend instead (if easy to do so)
- or just code your own backend - doing that in a test is an intended usage.

For the latter I mean similarly to how some tests in pubtools-pulp do it today:

```python
    fake_items = [
        FilePushItem(name="file1", dest=["some-repo", "other-repo", "/some/path"]),
        FilePushItem(name="file2", dest=["/some/path", "/other/path"]),
        FilePushItem(name="file3", dest=["final-repo"]),
    ]
    Source.register_backend("fake", lambda: fake_items)

    # ... then later pass 'fake:' URI to the code under test
```

One of the reasons pushsource has an extensible API is so that you can write tests like the above which simply hardcode whatever push items you want and avoid a dependency on the implementation details of any specific backend.

rbikar replied:
good suggestion, I'll rewrite that test data setup.

rohanpm commented on tests/rhsm/test_rhsm_client.py:
```diff
@@ -132,3 +132,48 @@ def test_create_image(requests_mocker, caplog):
+
+    # there should be 3 calls, last won't get any data, so we stop requesting another page.
+    assert m_list_images.call_count == 3
+    for req_history, offset in zip(m_list_images.request_history, [0, 1000, 2000]):
```
I feel like the result here in this test demonstrates what I was saying earlier that it seems wrong to always increase the offset by 1000.

If I understand right what the test simulates:

- we request with offset=0 and get 1000 items back
- then we request with offset=1000 and get 1 more item back
- then we request with offset=2000 and get 0 items back

The last request doesn't seem to make sense to me, if the server only provided us 1001 items so far, isn't it odd to make the next request starting at item number 2000 ?

rohanpm commented:
> this does seem to hang indefinitely:
> 
> ```
> with Executors.thread_pool().with_retry() as ex:
>    f = ex.submit()
>    f.result()
> ```

I would be interested in looking into that if you have a simple reproducer.

rbikar commented:
> > this does seem to hang indefinitely:
> > ```
> > with Executors.thread_pool().with_retry() as ex:
> >    f = ex.submit()
> >    f.result()
> > ```
> 
> I would be interested in looking into that if you have a simple reproducer.

Actually I accidentally had `f.result()` outside the `with` block so I guess this case is invalid when the executor has already been shutdown. Not sure what the expected behavior should be in this case (I guess ideally Future should fail be canceled immediately at the time executor is shut down.).

rbikar commented on pubtools/_ami/rhsm.py:
```diff
@@ -152,3 +159,27 @@ def create_image(
+            )
+            resp = resp_f.result().json()
+            if resp["pagination"]["count"]:
+                offset += 1000
```
Actually, docs of the service doesn't really say that. But there is `"count": 345` field in the response which can be used for this purposed.

rbikar commented on pubtools/_ami/tasks/delete.py:
```diff
@@ -0,0 +1,239 @@
+        group.add_argument(
+            "--limit",
+            nargs="+",
+            help="Only remove the specified AMIs",
```
yes, I'll fix that.

rbikar commented on tests/logs/push/test_ami_delete/test_ami_delete[keep_snapshot_false].jsonl:
```diff
@@ -0,0 +1 @@
+
```
Yeah, I agree that this a bug and should be fixed. Will try to address that.

rbikar commented:
Moved to draft.
Finally I have time to address the issues, couldn't do that sooner because of ticket-heavy ticketmastering sprint.

rbikar commented:
This is ready for review:
* improved pages handling while requesting RHSM service
* added saving push items 
* added usage of `step` decorator
* using `fake` pushsource backend in tests

rohanpm commented on pubtools/_ami/tasks/delete.py:
```diff
@@ -0,0 +1,253 @@
+    """Deletes one or more Amazon Machine Images on AWS from specified sources.
+
+    This command gets the AMIs from specified source, checks the image presence
+    in metadta service (e.g. RHSM) and makes image invisible in this service.
```
```suggestion
    in metadata service (e.g. RHSM) and makes image invisible in this service.
```

rohanpm commented on pubtools/_ami/tasks/delete.py:
```diff
@@ -0,0 +1,253 @@
+
+    This command gets the AMIs from specified source, checks the image presence
+    in metadta service (e.g. RHSM) and makes image invisible in this service.
+    This is followed by deleltino of image and related snapshot in AWS.
```
```suggestion
    This is followed by deletion of image and related snapshot in AWS.
```

rbikar commented:
Applied suggestions on typos and squashed commits - carrying over +1.
